File: Frontend/src/lib/applications.ts
import { getAccessToken } from "@/lib/auth";
import { getApiBaseUrl } from "@/lib/api";
import type { BocraLicenceType } from "@/lib/constants";

export type ApplicationStatus =
  | "draft"
  | "submitted"
  | "under_review"
  | "approved"
  | "rejected"
  | "requires_action";

export interface ApplicationListItem {
  id: string;
  reference_number: string;
  licence_type: string;
  status: ApplicationStatus;
  submitted_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface ApplicationDetail extends ApplicationListItem {
  applicant_id: string;
  form_data_a: Record<string, unknown> | null;
  form_data_b: Record<string, unknown> | null;
  form_data_c: Record<string, unknown> | null;
  form_data_d: Record<string, unknown> | null;
  documents: Array<{
    id: string;
    file_name: string;
    file_type: string | null;
    file_size: number | null;
    category: string | null;
    created_at: string;
  }>;
  admin_notes: string | null;
  decision_reason: string | null;
  decided_by: string | null;
  decided_at: string | null;
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const token = getAccessToken();
  const response = await fetch(`${getApiBaseUrl()}${path}`, {
    ...init,
    headers: {
      "Content-Type": "application/json",
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
      ...(init?.headers || {}),
    },
  });

  if (!response.ok) {
    let message = "Request failed";
    try {
      const body = await response.json();
      message = body.detail || body.message || message;
    } catch {
      // ignore parse failures
    }
    throw new Error(message);
  }

  return (await response.json()) as T;
}

export async function listApplications(): Promise<ApplicationListItem[]> {
  return request<ApplicationListItem[]>("/api/applications", { method: "GET" });
}

export async function createApplication(payload: {
  licence_type: BocraLicenceType;
  form_data_a?: Record<string, unknown>;
  form_data_b?: Record<string, unknown>;
  form_data_c?: Record<string, unknown>;
  form_data_d?: Record<string, unknown>;
}): Promise<ApplicationDetail> {
  return request<ApplicationDetail>("/api/applications", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function getApplication(applicationId: string): Promise<ApplicationDetail> {
  return request<ApplicationDetail>(`/api/applications/${applicationId}`, { method: "GET" });
}

export async function updateApplication(
  applicationId: string,
  payload: {
    licence_type?: BocraLicenceType;
    form_data_a?: Record<string, unknown>;
    form_data_b?: Record<string, unknown>;
    form_data_c?: Record<string, unknown>;
    form_data_d?: Record<string, unknown>;
  },
): Promise<ApplicationDetail> {
  return request<ApplicationDetail>(`/api/applications/${applicationId}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export async function submitApplication(applicationId: string): Promise<ApplicationDetail> {
  return request<ApplicationDetail>(`/api/applications/${applicationId}/submit`, {
    method: "POST",
    body: JSON.stringify({}),
  });
}

export async function updateApplicationStatus(
  applicationId: string,
  status: ApplicationStatus,
  adminNotes?: string,
): Promise<ApplicationDetail> {
  return request<ApplicationDetail>(`/api/applications/${applicationId}`, {
    method: "PATCH",
    body: JSON.stringify({
      status,
      admin_notes: adminNotes,
    }),
  });
}

File: backend/app/models/applications.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, StrictStr

from app.constants import BOCRA_LICENCE_TYPES


class ApplicationCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    licence_type: Literal[
        "Aircraft Radio Licence",
        "Amateur Radio License",
        "Broadcasting Licence",
        "Cellular Licence",
        "Citizen Band Radio Licence",
        "Point-to-Multipoint Licence",
        "Point-to-Point Licence",
        "Private Radio Communication Licence",
        "Radio Dealers Licence",
        "Radio Frequency Licence",
        "Satellite Service Licence",
        "Type Approval Licence",
        "VANS Licence",
    ]
    form_data_a: dict[str, Any] | None = None
    form_data_b: dict[str, Any] | None = None
    form_data_c: dict[str, Any] | None = None
    form_data_d: dict[str, Any] | None = None


class ApplicationPatchRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    licence_type: Literal[
        "Aircraft Radio Licence",
        "Amateur Radio License",
        "Broadcasting Licence",
        "Cellular Licence",
        "Citizen Band Radio Licence",
        "Point-to-Multipoint Licence",
        "Point-to-Point Licence",
        "Private Radio Communication Licence",
        "Radio Dealers Licence",
        "Radio Frequency Licence",
        "Satellite Service Licence",
        "Type Approval Licence",
        "VANS Licence",
    ] | None = None
    form_data_a: dict[str, Any] | None = None
    form_data_b: dict[str, Any] | None = None
    form_data_c: dict[str, Any] | None = None
    form_data_d: dict[str, Any] | None = None
    # Admin-only fields (will be validated in endpoint)
    status: Literal["draft", "submitted", "under_review", "waiting_for_payment", "approved", "rejected", "requires_action"] | None = None
    admin_notes: str | None = None
    decision_reason: str | None = None


class ApplicationRequestInfoRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    note: StrictStr = Field(min_length=1, max_length=2000)


class ApplicationDecideRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    status: Literal["approved", "rejected"]
    decision_reason: StrictStr = Field(min_length=1, max_length=1000)
    admin_notes: str | None = None


class ApplicationSubmitRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    pass  # No body required for submit


class ApplicationListItem(BaseModel):
    id: UUID
    reference_number: str
    licence_type: str
    status: str
    submitted_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class ApplicationDetail(BaseModel):
    id: UUID
    applicant_id: UUID
    reference_number: str
    licence_type: str
    status: str
    form_data_a: dict[str, Any] | None = None
    form_data_b: dict[str, Any] | None = None
    form_data_c: dict[str, Any] | None = None
    form_data_d: dict[str, Any] | None = None
    admin_notes: str | None = None
    decision_reason: str | None = None
    decided_by: UUID | None = None
    decided_at: datetime | None = None
    submitted_at: datetime | None = None
    documents: list["ApplicationDocumentSummary"] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class ApplicationDocumentSummary(BaseModel):
    id: UUID
    file_name: str
    file_type: str | None = None
    file_size: int | None = None
    category: str | None = None
    created_at: datetime


class ApplicationStatusLog(BaseModel):
    id: UUID
    old_status: str | None = None
    new_status: str
    changed_by: UUID | None = None
    reason: str | None = None
    created_at: datetime


class LicenceVerificationItem(BaseModel):
    licence_number: str
    customer_name: str
    licence_type: str
    issue_date: datetime
    expiration_date: datetime
    status: Literal["Active", "Expired", "Suspended"]


class LicenceVerificationResponse(BaseModel):
    items: list[LicenceVerificationItem]
    count: int
